### ISSUE TITLE
[Ops] 모델 관리와 연구자 간의 커뮤니케이션을 지원할 인프라 구축

### ISSUE THREAD

**Issue by robert-min** [FLYAI4/mlops-utils#1]
### 📌 Description
현재 각 연구자들이 개발한 모델의 성능을 비교하기 위해 직접 `.ipynb` 파일 또는 결과 그래프 이미지 및 수치를 직접 공유하는 등 많은 불편함과 어려움이 존재합니다.
모델의 성능 비교 및 관리를 할 수 있는 인프라를 개발하여 커뮤니케이션 향상에 기여합니다.

<br>

### 🎈 Goal
<p>$\tiny{구체적인\ 산출물을\ 포함한\ 목표를\ 작성해주세요.}$</p>

- 오픈소스 결정
  - 모델 관리를 할 수 있는 오픈소스 리서치
  - 비용, 팀 상황을 고려하여 오픈소스 결정
- 인프라 구축
  - 모델 관리 인프라 구축 진행
  - 운영 비용 확인 및 정리
  - 백업 및 복구 전략 설정 및 정리
- utils 개발
  - 활용성을 높이기 위한 utils 패키지 개발
  - pypi 패키지 공유(mlops-utils)
- 기술 공유
  - 패키지 사용 문서화
  - 기술 세미나

<br>

### ✏️ Todo
<p>$\tiny{목표\ 달성을\ 위해\ 해야할\ 일을\ 세부적으로\ 작성해주세요.}$</p>

- [x] 오픈소스 결정 📅 2023-12-27
  - [x] Kubeflow 리서치 및 문서화
  - [x] MLflow 리서치 및 문서화
  - [x] 비용, 팀 상황을 고려하여 오픈소스 결정 및 문서화
- [x] 인프라 구축 📅 2024-01-02
  - [x] 운영 방식 결정 및 문서화
  - [x] 운영 비용 정리 및 문서화
  - [x] 백업 및 복구 전략 결정 및 문서화
  - [x] 인프라 구축
- [x] utils 개발 📅 2024-01-02
  - [x] sklearn 로깅 기능 개발
  - [x] pytorch 로깅 기능 개발 
  - [x] figure 저장 기능 개발
- [x] 기술 공유  📅 2024-01-09

------

**Comment by robert-min** [comment]
# 오픈소스 결정
## 1. Kubeflow [링크](https://www.kubeflow.org/)
- 쿠버네티스 기반 AI 플랫폼으로 머신러닝 워크플로우를 자동화
  *Machine learing workflow : ML 모델을 개발하는 모든 과정을 의미. 데이터 전처리-EDA-하이퍼파라미터튜닝-머신러닝 모델 배포
- 머신러닝 워크 플로우 각 단계를 개발하는 데 필요한 도구를 구축하는 데 수작업을 최소화하고 AI 모델을 쉽게 개발하는 기능을 제공

### a. 제공하는 기능
머신러닝 모델의 전체 과정에 필요한 기능을 제공
- Experimental phase
	- 모델 구현
		- pytorch
		- scikit-learn
		- Tensorflow
	- 모델 학습
		- jupyter Notebook
		- Fairing
		- Pipelines(KFP) : 쿠버네티스 학습 파이프라인
	- 하이퍼파라미터 튜닝
		- Katib : 하이퍼파리미터 튜닝 담당(현재는 베타버전)
- production phase
	- 모델 서빙
		- KFServing : ML Serving을 담당(현재는 베타버전)
		- Pytorch
		- TFServing
		- Seldon
	- 모델 모니터링
		- TensorBoard
		- Metadata
- Central Dashboard
	- UI로 Kubeflow 서비스 이용, 리소스 제어
	- Notebook Service (ex. jupyter notebook)
		- stateless 상태임으로 작업 내용을 보존하지 않는 문제가 있음
		- k8s persistant volume 기능을 활용해야 함
	- CPU, Memory, GPU 쿠버네티스 클러스터가 지원하는 하에서 사용 가능
- Training Operator
- Multi Tenancy
- user 관리, usergroup 관리
- 인증 : Istio 
- 인가 : Istio + k8s

### b. 장단점
- 장점
  - k8s에서 모델을 훈련-배포 등 확장 가능한 방법을 제공
  - 병렬, 순차(DAG) 작업을 고려한 ML파이프라인 구축에 적합 -> AutoML
  - k8s 인프라를 관리하여 필요한 리소스를 활용하여 맞춤형 ML 솔루션을 제공
- 단점
  - k8s에 대한 이해가 필요
  - ML 서비스의 전체 과정을 플랫폼으로 제공할 대규모 팀에 적합

## 2. MLFlow [링크](https://mlflow.org/)
- ML의 라이프 사이클을 관리
- ML 모델의 실험을 tracking하고 model을 공유 및 deploy 할 수 있도록 지원하는 라이브러리

### a. 제공하는 기능
- Model tracking : ML 모델을 학습시킬 때 생기는 각종 파라미터, metric 결과 등을 logging하고 그 결과를 web ui에서 확인
- MLflow Projects : anaconda, docker 등을 사용해서 만들어 둔 모델을 reproducible하고 실행할 수 있도록 코드 패키지 형태로 제공
- MLflow Models : 동일한 모델을 Docker, Spark, AWS 등에 쉽게 배치할 수 있도록 지원
- MLflow Model Registry : MLflow 모델의 전체 라이프사이클을 공동으로 관리하기 위한 기능 제공


### b. 장단점
- 장점
  - MLFlow는 실험 및 버전 관리 모델을 추적하기 위한 Python 프로그램
  - 즉 인프라에 제한 없이 어느 환경에서나 실행할 수 있고, 단일 서비스로 운영되어 사용하기가 쉬움
- 단점
  - ML 모델 트레킹에 중점된 기능을 제공하기 때문에 Kubeflow와 같이 전체 AI 플랫폼으로 사용되기에 어려움이 있음(별도의 오픈소스와 함께 사용할 필요가 있음)

## 3. 오픈소스 결정

> - Kubeflow의 경우 쿠버네티스 환경을 구축과 운영 상이 리소스 관리 측면에서 얼려움이 있음
> - 추가로 프로젝트 비용을 고려했을 때 충분한 리소스 활용이 어려울 것으로 판단해 Kubeflow를 사용하기 보단 MLFlow를 기본으로 사용하고 필요한 기술을 붙여서 사용하는 방법으로 진행

- MLFlow 기반의 모델 tracking 오픈소스 사용

**Comment by robert-min** [comment]
# 인프라 구축
## 1. 운영 방식 결정
- ECS (or Container) 형태로 운영 or EC2에서 mlflow 운영
  - Dev 환경에서 model tracking을 위해 사용되기 때문에 Production 환경(k8s)과 분리되어서 사용
 - 모델 백업 전략
   - 모델 결과를 별도의 DB에 저장 -> S3?
     - 신규 모델(mlruns, mlartifacts)을 압축
   - Amazon EBS 스냅샷을 사용해 별도 저장
- 비용
  - ECS 비용
    - AWS 리소스(예: Amazon Elastic Cloud Compute(Amazon EC2) 인스턴스 또는 Amazon Elastic Block Store(Amazon EBS) 볼륨)에 대한 비용을 지불

**Comment by robert-min** [comment]
# MLflow 인프라 구축

## 1. 설치
```sh
# python 설치
sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt-get update

## package 업데이트 되었는지 확인
apt list | grep python3.10

sudo apt install python3.10

## 설치 확인
python3 --version

## 경로 지정
mkdir ~/mlflow && cd ~/mlflow

## 가상환경 설치
sudo apt install python3.8-venv
python3 -m venv .venv

## 가상환경 실행
source .venv/bin/activate

# mlflow 설치
pip3 install mlflow

# mlflow 실행
## host 지정해야만 외부에 접속할 수 있는 IP 열림
mlflow ui --host 0.0.0.0
```



## 2. mlflow ui 접속을 위한 인바인드 규칙 수정
- Azure에서 인바운드 규칙 5000번 열여주기!!

## 3. mlflow 간단한 실행

```python
import mlflow
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline

experiment_name = "minjun_researcher"
run_name = "20240116-sklearn"

mlflow.set_tracking_uri("{url}:5000")
mlflow.set_experiment(experiment_name)
mlflow.sklearn.autolog()

with mlflow.start_run(run_name=run_name) as run:
    noise = np.random.rand(100, 1)
    X = sorted(10 * np.random.rand(100, 1)) + noise
    y = sorted(10 * np.random.rand(100))

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)
    pipe = Pipeline([("scaler", StandardScaler()), ("lr", LinearRegression())])
    pipe.fit(X_train, y_train)
    preds = pipe.predict(X_test)
    print(preds)

```


## 4. mlflow 활용 라이브러리 

- [pypi 링크](https://pypi.org/project/mlops-utils/)

```sh
!pip install mlops-utils
```

## 예시 코드
```python
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from mlops_utils.mlflow_logger.sklearn import Sklearn

# URL 주소 변경
sklearn = Sklearn("{url}", 5000)
experiment_name = "minjun_researcher"
run_name = "20240116-sklearn"


@sklearn.logger(experiment_name, run_name)
def ml_run():
    noise = np.random.rand(100, 1)
    X = sorted(10 * np.random.rand(100, 1)) + noise
    y = sorted(10 * np.random.rand(100))

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)
    pipe = Pipeline([("scaler", StandardScaler()), ("lr", LinearRegression())])
    pipe.fit(X_train, y_train)
    preds = pipe.predict(X_test)
    print(preds)

    sklearn.post_board(y_test, preds)

ml_run()

```
